### ISSUE TITLE
Unable to install "@tableau/embedding-api" using yarn

### ISSUE THREAD

**Issue by oliqwe** [tableau/embedding-api-v3-guide#67]
Hello,

for some reason when installing package with yarn it does not find some of the packages.
i created a clean project with no dependencies and used  ` yarn add @tableau/embedding-api`

$ yarn add @tableau/embedding-api --verbose
yarn add v1.22.19
verbose 0.4657057 current time: 2024-04-03T06:37:04.188Z
[1/4] Resolving packages...
verbose 0.7773877 Performing "GET" request to "https://registry.yarnpkg.com/@tableau%2fembedding-api".
verbose 2.4755437 Request "https://registry.yarnpkg.com/@tableau%2fembedding-api" finished with status code 200.
verbose 2.477336 Performing "GET" request to "https://registry.yarnpkg.com/@tableau%2fapi-external-contract-js".
verbose 2.478122 Performing "GET" request to "https://registry.yarnpkg.com/@tableau%2fapi-internal-contract-js".
verbose 2.4787556 Performing "GET" request to "https://registry.yarnpkg.com/@tableau%2fapi-shared-js".
verbose 3.2612582 Request "https://registry.yarnpkg.com/@tableau%2fapi-shared-js" finished with status code 404.
verbose 3.2924577 Request "https://registry.yarnpkg.com/@tableau%2fapi-internal-contract-js" finished with status code 404.
verbose 3.3581252 Request "https://registry.yarnpkg.com/@tableau%2fapi-external-contract-js" finished with status code 404.
verbose 3.3659601 Error: Couldn't find package "@tableau/api-external-contract-js@1.189.0" required by "@tableau/embedding-api" on the "npm" registry.
    at MessageError.ExtendableBuiltin (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:721:66)
    at new MessageError (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:750:123)
    at PackageRequest.<anonymous> (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:36590:17)
    at Generator.throw (<anonymous>)
    at step (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:310:30)
    at C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:323:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
error Couldn't find package "@tableau/api-external-contract-js@1.189.0" required by "@tableau/embedding-api" on the "npm" registry.
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
Error: Couldn't find package "@tableau/api-shared-js@1.231.0" required by "@tableau/embedding-api" on the "npm" registry.
    at MessageError.ExtendableBuiltin (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:721:66)
    at new MessageError (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:750:123)
    at PackageRequest.<anonymous> (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:36590:17)
    at Generator.throw (<anonymous>)
    at step (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:310:30)
    at C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:323:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Error: Couldn't find package "@tableau/api-internal-contract-js@1.162.0" required by "@tableau/embedding-api" on the "npm" registry.
    at MessageError.ExtendableBuiltin (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:721:66)
    at new MessageError (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:750:123)
    at PackageRequest.<anonymous> (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:36590:17)
    at Generator.throw (<anonymous>)
    at step (C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:310:30)
    at C:\Users\olegm\AppData\Roaming\npm\node_modules\yarn\lib\cli.js:323:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

**Comment by MarkN3DTI** [comment]
@oliqwe I had this same problem and managed to find a workaround. First I had to use `npm` (version 10.5.0) to do an initial install of the embedding API:

`npm i @tableau/embedding-api`

The "private" libraries are bundled with the base package, so once the base package was installed, I added a `resolutions` section to my `package.json` file:

`  "resolutions": {
    "@tableau/api-external-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-external-contract-js",
    "@tableau/api-internal-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-internal-contract-js",
    "@tableau/api-shared-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-shared-js"
  },`

Now I can run `yarn install` to maintain my other packages without hitting the permissions/404 error with the private libraries.

**Comment by arthurgailes** [comment]
Same problem with pnpm: 
```
 ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@tableau%2Fapi-internal-contract-js: Not Found - 404

This error happened while installing the dependencies of @tableau/embedding-api@3.10.0

@tableau/api-internal-contract-js is not in the npm registry, or you have no permission to fetch it.

No authorization header was set for the request.
```

Given a monorepo setup, the package.json snippet looks like:
```
"pnpm": {
    "overrides": {
      "@tableau/api-external-contract-js": "file:./libs/shared/feature/tableau/node_modules/.ignored/@tableau/embedding-api/node_modules/@tableau/api-external-contract-js",
      "@tableau/api-internal-contract-js": "file:./libs/shared/feature/tableau/node_modules/.ignored/@tableau/embedding-api/node_modules/@tableau/api-internal-contract-js",
      "@tableau/api-shared-js": "file:./libs/shared/feature/tableau/node_modules/.ignored/@tableau/embedding-api/node_modules/@tableau/api-shared-js"
    }
  }
```

**Comment by anyoung-tableau** [comment]
This is because yarn 2 does not support bundle dependencies, which the package currently leverages to include those internal packages. https://yarnpkg.com/migration/guide#dont-use-bundledependencies

Team is aware of the issue but no ETA to share yet.

**Comment by learnercys** [comment]
similar for me with npm

```
npm error 404  '@tableau/api-external-contract-js@1.173.0' is not in this registry.
```

does someone have any ETA?

**Comment by anyoung-tableau** [comment]
`@tableau/api-external-contract-js`, `@tableau/api-internal-contract-js`, and `@tableau/api-shared-js` are bundled with `@tableau/embedding-api`. You don't need to try to install them by themselves.

`npm install @tableau/embedding-api` should install everything you need.

**Comment by learnercys** [comment]
@anyoung-tableau as you suggested I am only installing embedding-api, I made it work with npm overrides:

```
"overrides": {
    "@tableau/api-external-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-external-contract-js",
    "@tableau/api-internal-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-internal-contract-js",
    "@tableau/api-shared-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-shared-js"
  }
```

unfourtunately this is not working with my tests. Also I am using a private registry in case this helps

**Comment by karthickmerck** [comment]
Any solution for this resolution works in local but when use azure pipelines getting issue.

 "resolutions": {
    "@tableau/api-external-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-external-contract-js",
    "@tableau/api-internal-contract-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-internal-contract-js",
    "@tableau/api-shared-js": "file:./node_modules/@tableau/embedding-api/node_modules/@tableau/api-shared-js"
  },

error Package "" refers to a non-existing file

**Comment by Miguel-A-Jara** [comment]
Hello, does anybody have any updates on this? I wonder if there are any other workarounds?

**Comment by anyoung-tableau** [comment]
Hey folks, I believe I have a fix for this. I've published a temporary, fixed package to [@tableau/embedding-api-alpha](https://www.npmjs.com/package/@tableau/embedding-api-alpha) and verified it locally with npm, yarn, pnpm, bun, and deno. It would be great if you could install it in your dev environment to see if you are unblocked. Please note that it is a pre-release version of the Embedding API specifically made to test a fix for this issue. There's no support for it otherwise.

Also, if you are using a version older than 3.12 of `@tableau/embedding-api` please let me know so I can know which old versions I should backport this fix to. Otherwise, I'll try to publish a 3.12.1 patch tomorrow.

**Comment by amiteshksharma** [comment]
> Hey folks, I believe I have a fix for this. I've published a temporary, fixed package to [@tableau/embedding-api-alpha](https://www.npmjs.com/package/@tableau/embedding-api-alpha) and verified it locally with npm, yarn, pnpm, bun, and deno. It would be great if you could install it in your dev environment to see if you are unblocked. Please note that it is a pre-release version of the Embedding API specifically made to test a fix for this issue. There's no support for it otherwise.
> 
> Also, if you are using a version older than 3.12 of `@tableau/embedding-api` please let me know so I can know which old versions I should backport this fix to. Otherwise, I'll try to publish a 3.12.1 patch tomorrow.

Is there plans to fix this bug for version 3.10? Currently have this issue when trying to download for our build process and none of the solutions above are working for me. Do you know if there will be a fix for it soon?

**Comment by anyoung-tableau** [comment]
@amiteshksharma yeah, I will backport it to 3.10. Were you able to confirm you are able to install the alpha version in your build process? I want to make sure the fix works for folks before I publish the new NPM package, including to old versions.

Thanks,
Andy

**Comment by amiteshksharma** [comment]
@anyoung-tableau  I run into a build issue with npm build

`13:14:25
+ cd ./
13:14:25
+ npm run build
13:14:25
> activity-web@0.1.0 build
13:14:25
> CI=false react-scripts build
13:14:26
Creating an optimized production build...
13:14:35
Failed to compile.
13:14:35
Module not found: Error: Can't resolve '@tableau/api-external-contract-js' in '/mnt/looper/workspace/StoreSystemsMobility/Amp2.0-CMS/pr-1026/ws/node_modules/@tableau/embedding-api-alpha/lib/src'` 

Not sure why is occurs and not sure how to resolve it, do I need to specify anything in my package.json?

**Comment by anyoung-tableau** [comment]
To confirm, you've removed [@tableau/embedding-api](https://www.npmjs.com/package/@tableau/embedding-api) from your project, and added [@tableau/embedding-api-alpha](https://www.npmjs.com/package/@tableau/embedding-api-alpha)? I'm not certain what you're seeing is the same issue as #67.

**Comment by amiteshksharma** [comment]
Update, it works locally with this alpha version, but when I run it through my company's build pipeline, it fails at the build step saying it can't find one of the modules... 

Should I consider using the <script> option to import the tableu embed feature for my react application? I'm not sure how to resolve this build issue then since it can't find the internal dependencies

**Comment by anyoung-tableau** [comment]
Thanks @amiteshksharma , since issue #67 is only for tracking the inability to _install_ the NPM package with yarn, this sounds like a different issue. If you could, please debug further until you have enough information for the issue to be reproducible on our end and open a new issue (e.g. version of npm, components of build process, etc). To start with, make sure that after the `npm install` your node_modules should contain the following structure:

```
+---node_modules
|   +---@tableau
|   |   \---embedding-api
|   |       \---node_modules
|   |           +---@tableau
|   |           |   +---api-external-contract-js
|   |           |   +---api-internal-contract-js
|   |           |   \---api-shared-js
```

**Comment by anyoung-tableau** [comment]
I have published patched versions of 3.10, 3.11, and 3.12 that should now install using Yarn without error. Please let me know within the next few days if you still are having issues installing.

**Comment by radsar** [comment]
> I have published patched versions of 3.10, 3.11, and 3.12 that should now install using Yarn without error. Please let me know within the next few days if you still are having issues installing.

@anyoung-tableau Could you please publish the same for 3.5 version? We are also facing the same issue while installing @tableau/embedding-api@3.5.0 using yarn. It would really helpful if we could patch it for 3.5. Thanks in advance

**Comment by karthickmerck** [comment]
error Couldn't find package "@tableau/api-external-contract-js@1.211.0" required by "@tableau/embedding-api@^3.12.0" on the "npm" registry.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: Couldn't find package "@tableau/api-shared-js@1.254.0" required by "@tableau/embedding-api@^3.12.0" Still i am getting this issue

**Comment by karthickmerck** [comment]
> I have published patched versions of 3.10, 3.11, and 3.12 that should now install using Yarn without error. Please let me know within the next few days if you still are having issues installing.

Still getting same issue

**Comment by radsar** [comment]
> > I have published patched versions of 3.10, 3.11, and 3.12 that should now install using Yarn without error. Please let me know within the next few days if you still are having issues installing.
> 
> Still getting same issue

@karthickmerck have you removed [@tableau/embedding-api](https://www.npmjs.com/package/@tableau/embedding-api) from your project, and added [@tableau/embedding-api-alpha](https://www.npmjs.com/package/@tableau/embedding-api-alpha)?

**Comment by dima-sasg** [comment]
Same issue for 3.12.0 version @anyoung-tableau . Blocker.

**Comment by d-errecalde** [comment]
For the [@tableau/embedding-api@3.13.0](https://www.npmjs.com/package/@tableau/embedding-api) the dependancies (@tableau/api-shared-js, @tableau/api-external-contract-js and @tableau/api-internal-contract-js) are removed by yarn or npm   executing npm install or yarn install to follow the new practices rules. So @anyoung-tableau the point is to not have children dependancies inside @tableau/embedding-api but instead only sibling dependancies should fix the issue if you can try this on next version. Many thanks in advance !

+---node_modules
|   +---@tableau
|   |   \----embedding-api
|   |   \----api-external-contract-js
|   |   \----api-internal-contract-js
|   |   \----api-shared-js
